File: frontend/src/api/macro.ts
import { api } from "./client"

export interface MacroRecord {
  id: number
  symbol: string
  data_source: string
  date: string
  value: number
  data: Record<string, any> | null
  created_at: string
  updated_at: string
}

export interface MacroSymbol {
  symbol: string
  data_source: string
}

export const macroApi = {
  listSymbols: () => api.get<MacroSymbol[]>("/macro/symbols"),

  query: (
    symbol?: string,
    startDate?: string,
    endDate?: string,
    limit?: number,
  ) => {
    const params = new URLSearchParams()
    if (symbol) params.set("symbol", symbol)
    if (startDate) params.set("start_date", startDate)
    if (endDate) params.set("end_date", endDate)
    if (limit) params.set("limit", String(limit))
    return api.get<MacroRecord[]>(`/macro/?${params}`)
  },
}

File: backend/src/farseer/schemas/macro.py
"""Macro-economic data schemas for API serialization."""

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, field_validator

from farseer.sources.datasource import DataSource


class MacroBase(BaseModel):
    symbol: str
    data_source: DataSource = DataSource.akshare
    date: date
    value: float
    data: dict[str, Any] | None = None

    @field_validator("data", mode="before")
    @classmethod
    def coerce_data(cls, v: Any) -> Any:
        if isinstance(v, str):
            import json
            return json.loads(v)
        return v


class MacroOut(MacroBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MacroQuery(BaseModel):
    symbol: str | None = None
    data_source: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    limit: int = 200
